Forta-Bots-Challenge-4/liquidatable-positions/src/agent.ts

```diff
@@ -0,0 +1,107 @@
+  COMET_ABI,
+  FETCHER_CONFIG_USDC,
+  FETCHER_CONFIG_WETH,
+} from "./constants";
+import { providers } from "ethers";
+import PositionFetcher from "./position.fetcher";
+import { createNewPositionFinding, createLiquidatableFinding } from "./finding";
+import { PositionDataResponse, emptyPositionData, FetcherConfig, Liquidatable } from "./agent.utils";
+
+let positionFetcherUSDC: PositionFetcher;
+let positionFetcherWETH: PositionFetcher;
+
+export const provideInitialize = (
+  provider: providers.Provider,
+  fetcherConfigUSDC: FetcherConfig,
+  fetcherConfigWETH: FetcherConfig,
+  erc20ABI: string[],
+  cometABI: string[],
+  borrowCallNames: string[]
+): Initialize => {
+  return async () => {
+    positionFetcherUSDC = new PositionFetcher(provider, fetcherConfigUSDC, erc20ABI, cometABI, borrowCallNames);
+    positionFetcherWETH = new PositionFetcher(provider, fetcherConfigWETH, erc20ABI, cometABI, borrowCallNames);
+  };
+};
+
+export const provideHandleTransaction = (cometWithdrawSupplySignatures: string[]): HandleTransaction => {
+  return async (txEvent: TransactionEvent) => {
+    const findings: Finding[] = [];
+    const baseAssetAddresses = [positionFetcherUSDC.baseAssetAddress, positionFetcherWETH.baseAssetAddress];
+
+    // filter transaction events for withdraw/supply calls
+    await Promise.all(
+      txEvent
+        .filterFunction(cometWithdrawSupplySignatures, [
+          positionFetcherUSDC.cometAddress,
+          positionFetcherWETH.cometAddress,
+        ])
+        .map(async (call) => {
+          if (txEvent.to) {
+            // dispatch withdraw/supply call info to base asset's fetcher
+            const positionResponse: PositionDataResponse =
+              call.address === positionFetcherUSDC.cometAddress
+                ? await positionFetcherUSDC.delegateCall(call, txEvent.from, txEvent.blockNumber)
+                : call.address === positionFetcherWETH.cometAddress
+                  ? await positionFetcherWETH.delegateCall(call, txEvent.from, txEvent.blockNumber)
+                  : { newPosition: false, position: emptyPositionData };
+
+            // generate findings for new or updated positions
+            if (positionResponse.newPosition) {
+              findings.push(createNewPositionFinding(baseAssetAddresses, positionResponse.position));
+            }
+          }
+        })
+    );
+    return findings;
+  };
+};
+
+export const provideHandleBlock = (): HandleBlock => {
+  return async (BlockEvent: BlockEvent) => {
+    const findings: Finding[] = [];
+    const blockNumber = BlockEvent.blockNumber;
+    const baseAssetAddresses = [positionFetcherUSDC.baseAssetAddress, positionFetcherWETH.baseAssetAddress];
+
+    // get liquidatable status from all positions currently in each base asset's cache
+    const positions: Liquidatable[][] = await Promise.all([
+      positionFetcherUSDC.monitorPositions(blockNumber),
+      positionFetcherWETH.monitorPositions(blockNumber),
+    ]);
```

Comment:
```suggestion
    const positions: Liquidatable[] = await Promise.all([
      ...positionFetcherUSDC.monitorPositions(blockNumber),
      ...positionFetcherWETH.monitorPositions(blockNumber),
    ]);
```
Suggestion to prevent the nested arrays, since the "position" array doesn't seem to be necessary in L86-90. Would also prevent the usage of two `.forEach()`.

Reply:
I think this was my original implementation but since monitorPositions returns a Promise<Liquidatable[]> it can't be deconstructed by the spread operator. I originally had monitorPositions returning a Promise<Liquidatable>[] but it threw a compiler error saying it's not a valid return type, I guess it would also be wrapped in another Promise anyways?